File: src/frontend/apps/admin-antd/src/api/system/product.ts
import type { Recordable } from '@vben/types';

import { requestClient } from '#/api/request';

/** 产品类型项（下拉、列表、维护） */
export interface ProductTypeItem {
  id: string;
  name: string;
  sortOrder: number;
  visible: boolean;
}

export namespace ProductApi {
  export interface ProductItem {
    id: string;
    productTypeId?: string;
    status?: boolean;
    name: string;
    code: string;
    model: string;
    unit: string;
    barcode?: string;
    activationCode?: string;
    priceStandard?: string;
    marketSales?: string;
    description?: string;
    costPrice?: number;
    customerPrice?: number;
    qty?: number;
    tags?: string;
    feature?: string;
    configuration?: string;
    instructions?: string;
    installProcess?: string;
    operationProcessResources?: string;
    introduction?: string;
    introductionResources?: string;
    imagePath?: string;
    categoryId?: string | null;
    /** 列表/详情查询联表返回 */
    categoryName?: string;
    supplierId?: string | null;
  }

  /** 创建/更新产品请求 */
  export interface ProductCreateUpdate {
    productTypeId: string;
    status: boolean;
    name: string;
    code: string;
    model: string;
    unit: string;
    barcode: string;
    activationCode: string;
    priceStandard: string;
    marketSales: string;
    description: string;
    costPrice: number;
    customerPrice: number;
    qty: number;
    tags: string;
    feature: string;
    configuration: string;
    instructions: string;
    installProcess: string;
    operationProcessResources: string;
    introduction: string;
    introductionResources: string;
    imagePath: string;
    categoryId?: string | null;
    supplierId?: string | null;
  }
}

/** 产品分类树节点（用于 TreeSelect） */
export interface ProductCategoryTreeItem {
  id: string;
  name: string;
  remark: string;
  parentId: string | null;
  sortOrder: number;
  visible: boolean;
  isDiscount: boolean;
  children: ProductCategoryTreeItem[];
}

/** 产品分类单条（用于编辑回填） */
export interface ProductCategoryItem {
  id: string;
  name: string;
  remark: string;
  parentId: string | null;
  sortOrder: number;
  visible: boolean;
  isDiscount: boolean;
}

/** 供应商项（用于下拉、列表） */
export interface SupplierItem {
  id: string;
  fullName: string;
  shortName: string;
  contact: string;
  phone: string;
  email?: string;
  address?: string;
  remark?: string;
}

/** 产品参数项 */
export interface ProductParameterItem {
  id: string;
  productId: string;
  year: string;
  description: string;
}

/**
 * 获取产品列表（分页，用于订单明细产品下拉及产品维护列表）
 */
async function getProductList(params: Recordable<any>) {
  const res = await requestClient.get<{
    items: ProductApi.ProductItem[];
    total: number;
  }>('/products', { params });
  return res;
}

/**
 * 获取产品详情（用于编辑回填）
 */
async function getProduct(id: string) {
  return requestClient.get<ProductApi.ProductItem>(`/products/${id}`);
}

/**
 * 创建产品
 */
async function createProduct(data: ProductApi.ProductCreateUpdate) {
  return requestClient.post<{ id: string }>('/products', data);
}

/**
 * 更新产品
 */
async function updateProduct(id: string, data: ProductApi.ProductCreateUpdate) {
  return requestClient.put(`/products/${id}`, data);
}

/**
 * 删除产品
 */
async function deleteProduct(id: string) {
  return requestClient.delete(`/products/${id}`);
}

/**
 * 获取产品类型列表（用于产品表单、产品类型维护）
 */
async function getProductTypeList(includeInvisible = false) {
  return requestClient.get<ProductTypeItem[]>('/product-types', {
    params: { includeInvisible },
  });
}

/**
 * 获取产品类型详情（用于编辑回填）
 */
async function getProductType(id: string) {
  return requestClient.get<ProductTypeItem>(`/product-types/${id}`);
}

/**
 * 创建产品类型
 */
async function createProductType(data: { name: string; sortOrder?: number; visible?: boolean }) {
  return requestClient.post<{ id: string }>('/product-types', data);
}

/**
 * 更新产品类型
 */
async function updateProductType(
  id: string,
  data: { name: string; sortOrder: number; visible: boolean },
) {
  return requestClient.put(`/product-types/${id}`, data);
}

/**
 * 删除产品类型
 */
async function deleteProductType(id: string) {
  return requestClient.delete(`/product-types/${id}`);
}

/**
 * 获取产品分类树（用于产品表单分类选择、分类管理）
 */
async function getProductCategoryTree(includeInvisible = false, cacheBust = false) {
  return requestClient.get<ProductCategoryTreeItem[]>('/product-categories/tree', {
    params: { includeInvisible, ...(cacheBust ? { _t: Date.now() } : {}) },
  });
}

/**
 * 获取产品分类详情（用于编辑回填）
 */
async function getProductCategory(id: string) {
  return requestClient.get<ProductCategoryItem>(`/product-categories/${id}`);
}

/**
 * 创建产品分类
 */
async function createProductCategory(data: {
  name: string;
  remark: string;
  parentId?: string | null;
  sortOrder?: number;
  visible?: boolean;
  isDiscount?: boolean;
}) {
  return requestClient.post<{ id: string }>('/product-categories', data);
}

/**
 * 更新产品分类
 */
async function updateProductCategory(
  id: string,
  data: {
    name: string;
    remark: string;
    parentId?: string | null;
    sortOrder?: number;
    visible?: boolean;
    isDiscount?: boolean;
  },
) {
  return requestClient.put(`/product-categories/${id}`, data);
}

/**
 * 删除产品分类
 */
async function deleteProductCategory(id: string) {
  return requestClient.delete(`/product-categories/${id}`);
}

/**
 * 获取供应商列表（用于产品表单供应商选择、供应商管理）
 */
async function getSupplierList(keyword?: string) {
  return requestClient.get<SupplierItem[]>('/suppliers', {
    params: { keyword: keyword ?? undefined },
  });
}

/**
 * 获取供应商详情（用于编辑回填）
 */
async function getSupplier(id: string) {
  return requestClient.get<SupplierItem>(`/suppliers/${id}`);
}

/**
 * 创建供应商
 */
async function createSupplier(data: {
  fullName: string;
  shortName: string;
  contact: string;
  phone: string;
  email?: string;
  address?: string;
  remark?: string;
}) {
  return requestClient.post<{ id: string }>('/suppliers', data);
}

/**
 * 更新供应商
 */
async function updateSupplier(
  id: string,
  data: {
    fullName: string;
    shortName: string;
    contact: string;
    phone: string;
    email?: string;
    address?: string;
    remark?: string;
  },
) {
  return requestClient.put(`/suppliers/${id}`, data);
}

/**
 * 删除供应商
 */
async function deleteSupplier(id: string) {
  return requestClient.delete(`/suppliers/${id}`);
}

/**
 * 获取某产品的参数列表
 */
async function getProductParameterList(productId: string) {
  return requestClient.get<ProductParameterItem[]>(`/products/${productId}/parameters`);
}

/**
 * 创建产品参数
 */
async function createProductParameter(
  productId: string,
  data: { year: string; description: string },
) {
  return requestClient.post<{ id: string }>(`/products/${productId}/parameters`, data);
}

/**
 * 更新产品参数
 */
async function updateProductParameter(
  id: string,
  data: { year: string; description: string },
) {
  return requestClient.put(`/product-parameters/${id}`, data);
}

/**
 * 删除产品参数
 */
async function deleteProductParameter(id: string) {
  return requestClient.delete(`/product-parameters/${id}`);
}

export {
  getProductList,
  getProduct,
  createProduct,
  updateProduct,
  deleteProduct,
  getProductTypeList,
  getProductType,
  createProductType,
  updateProductType,
  deleteProductType,
  getProductCategoryTree,
  getProductCategory,
  createProductCategory,
  updateProductCategory,
  deleteProductCategory,
  getSupplierList,
  getSupplier,
  createSupplier,
  updateSupplier,
  deleteSupplier,
  getProductParameterList,
  createProductParameter,
  updateProductParameter,
  deleteProductParameter,
};

File: src/Ncp.Admin.Domain/AggregatesModel/ProductAggregate/Product.cs
using Ncp.Admin.Domain;
using Ncp.Admin.Domain.AggregatesModel.ProductCategoryAggregate;
using Ncp.Admin.Domain.AggregatesModel.ProductTypeAggregate;
using Ncp.Admin.Domain.AggregatesModel.SupplierAggregate;

namespace Ncp.Admin.Domain.AggregatesModel.ProductAggregate;

/// <summary>
/// 产品 ID（强类型）
/// </summary>
public partial record ProductId : IGuidStronglyTypedId;

/// <summary>
/// 产品聚合根：供订单明细关联
/// </summary>
public class Product : Entity<ProductId>, IAggregateRoot
{
    /// <summary>EF/序列化用</summary>
    protected Product() { }

    /// <summary>产品类型 ID（关联产品类型聚合）</summary>
    public ProductTypeId ProductTypeId { get; private set; } = default!;

    /// <summary>状态（是否有效）</summary>
    public bool Status { get; private set; }

    /// <summary>产品名称</summary>
    public string Name { get; private set; } = string.Empty;

    /// <summary>产品编号/货号</summary>
    public string Code { get; private set; } = string.Empty;

    /// <summary>型号</summary>
    public string Model { get; private set; } = string.Empty;

    /// <summary>单位</summary>
    public string Unit { get; private set; } = string.Empty;

    /// <summary>条码</summary>
    public string Barcode { get; private set; } = string.Empty;

    /// <summary>激活码</summary>
    public string ActivationCode { get; private set; } = string.Empty;

    /// <summary>价格标准（文本）</summary>
    public string PriceStandard { get; private set; } = string.Empty;

    /// <summary>市场销售</summary>
    public string MarketSales { get; private set; } = string.Empty;

    /// <summary>描述</summary>
    public string Description { get; private set; } = string.Empty;

    /// <summary>成本价</summary>
    public decimal CostPrice { get; private set; }

    /// <summary>客户价</summary>
    public decimal CustomerPrice { get; private set; }

    /// <summary>库存数量</summary>
    public int Qty { get; private set; }

    /// <summary>标签</summary>
    public string Tags { get; private set; } = string.Empty;

    /// <summary>功能特点</summary>
    public string Feature { get; private set; } = string.Empty;

    /// <summary>硬件配置</summary>
    public string Configuration { get; private set; } = string.Empty;

    /// <summary>使用说明</summary>
    public string Instructions { get; private set; } = string.Empty;

    /// <summary>操作流程（原安装流程）</summary>
    public string InstallProcess { get; private set; } = string.Empty;

    /// <summary>操作流程资源（上传的资源存储 Key，JSON 数组字符串）</summary>
    public string OperationProcessResources { get; private set; } = string.Empty;

    /// <summary>产品介绍</summary>
    public string Introduction { get; private set; } = string.Empty;

    /// <summary>产品介绍资源（上传的资源存储 Key，JSON 数组字符串）</summary>
    public string IntroductionResources { get; private set; } = string.Empty;

    /// <summary>图片路径/照片存储 Key</summary>
    public string ImagePath { get; private set; } = string.Empty;

    /// <summary>产品分类 ID（无分类时由端点传 new ProductCategoryId(Guid.Empty)）</summary>
    public ProductCategoryId CategoryId { get; private set; } = default!;

    /// <summary>供应商 ID（无供应商时由端点传 new SupplierId(Guid.Empty)）</summary>
    public SupplierId SupplierId { get; private set; } = default!;

    /// <summary>
    /// 创建产品（必填项由端点赋默认值，聚合内不判空）
    /// </summary>
    public Product(
        ProductTypeId productTypeId,
        bool status,
        string name,
        string code,
        string model,
        string unit,
        string barcode,
        string activationCode,
        string priceStandard,
        string marketSales,
        string description,
        decimal costPrice,
        decimal customerPrice,
        int qty,
        string tags,
        string feature,
        string configuration,
        string instructions,
        string installProcess,
        string operationProcessResources,
        string introduction,
        string introductionResources,
        string imagePath,
        ProductCategoryId categoryId,
        SupplierId supplierId)
    {
        ProductTypeId = productTypeId;
        Status = status;
        Name = name;
        Code = code;
        Model = model;
        Unit = unit;
        Barcode = barcode;
        ActivationCode = activationCode ?? string.Empty;
        PriceStandard = priceStandard ?? string.Empty;
        MarketSales = marketSales ?? string.Empty;
        Description = description;
        CostPrice = costPrice;
        CustomerPrice = customerPrice;
        Qty = qty;
        Tags = tags;
        Feature = feature;
        Configuration = configuration;
        Instructions = instructions;
        InstallProcess = installProcess;
        OperationProcessResources = operationProcessResources ?? string.Empty;
        Introduction = introduction;
        IntroductionResources = introductionResources ?? string.Empty;
        ImagePath = imagePath;
        CategoryId = categoryId;
        SupplierId = supplierId;
    }

    /// <summary>
    /// 更新产品信息
    /// </summary>
    public void Update(
        ProductTypeId productTypeId,
        bool status,
        string name,
        string code,
        string model,
        string unit,
        string barcode,
        string activationCode,
        string priceStandard,
        string marketSales,
        string description,
        decimal costPrice,
        decimal customerPrice,
        int qty,
        string tags,
        string feature,
        string configuration,
        string instructions,
        string installProcess,
        string operationProcessResources,
        string introduction,
        string introductionResources,
        string imagePath,
        ProductCategoryId categoryId,
        SupplierId supplierId)
    {
        ProductTypeId = productTypeId;
        Status = status;
        Name = name;
        Code = code;
        Model = model;
        Unit = unit;
        Barcode = barcode;
        ActivationCode = activationCode ?? string.Empty;
        PriceStandard = priceStandard ?? string.Empty;
        MarketSales = marketSales ?? string.Empty;
        Description = description;
        CostPrice = costPrice;
        CustomerPrice = customerPrice;
        Qty = qty;
        Tags = tags;
        Feature = feature;
        Configuration = configuration;
        Instructions = instructions;
        InstallProcess = installProcess;
        OperationProcessResources = operationProcessResources ?? string.Empty;
        Introduction = introduction;
        IntroductionResources = introductionResources ?? string.Empty;
        ImagePath = imagePath;
        CategoryId = categoryId;
        SupplierId = supplierId;
    }
}
